internal/rules/engine.go

```diff
@@ -92,393 +80,137 @@ func NewInMemoryEngineWithNamespaces(
 	return engine
 }
 
-func (i *InMemoryEngine) Reconcile(ctx context.Context) error {
-	i.mu.Lock()
-	defer i.mu.Unlock()
-
-	wsaEnumerable, err := i.GetWorkloadServiceAccounts(ctx)
-	if err != nil {
-		return err
-	}
+func (i *InMemoryEngine) GetTargetNamespaces() []string {
+	return i.targetNamespaces
+}
 
-	// Fetch ClusterWorkloadServiceAccounts
-	cwsaEnumerable, err := i.GetClusterWorkloadServiceAccounts(ctx)
-	if err != nil {
-		return err
+func (i *InMemoryEngine) SetGCTracker(tracker GCTrackerInterface) {
+	if rms, ok := i.ResourceManagement.(interface{ SetGCTracker(GCTrackerInterface) }); ok {
+		rms.SetGCTracker(tracker)
 	}
+}
 
-	// Get our target namespaces
-	if i.lookupNamespaces {
-		targetNamespaces, err := i.DiscoverTargetNamespaces(ctx, i.client)
-		if err != nil {
-			return fmt.Errorf("failed to discover target namespaces: %w", err)
-		}
-		i.targetNamespaces = targetNamespaces
-	}
+func (i *InMemoryEngine) GetOrDiscoverTargetNamespaces(ctx context.Context) ([]string, error) {
+	i.mu.RLock()
+	targetNamespaces := i.targetNamespaces
+	lookupNamespaces := i.lookupNamespaces
+	i.mu.RUnlock()
```

Comment:
I'm a bit unclear on the logic behind using `GetTargetNamespaces` vs `GetOrDiscoverTargetNamespaces` and why only one has the mutex around it.

Reply:
Liam goofed

We talked about this, `GetOrDiscoverTargetNamespaces` should be saving some values